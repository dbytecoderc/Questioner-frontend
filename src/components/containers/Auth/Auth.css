/* .error {
  background: red;
  color: white;
  padding: 5px 5px;
  font-size: 1rem;
  font-weight: 400;
  text-transform: uppercase;
  text-align: center;
  margin: .5rem 0;
  border-radius: 2px;
} */

.validation--requirement{
  color: #f00;
  margin: .2rem 0;
  font-size: 1rem;
}

.validation--requirement--last{
  margin: .2rem;
  color: #f00;
}

.success {
  background: green;
  color: white;
  padding: 5px 5px;
  font-size: 1rem;
  font-weight: 500;
  text-transform: uppercase;
  text-align: center;
  margin: .5rem 0;
  border-radius: 2px;
}

 .lds-ring {
  display: inline-block;
  position: relative;
  width: 25px;
  height: 25px;
}
.lds-ring div {
  box-sizing: border-box;
  display: block;
  position: absolute;
  width: 30px;
  height: 30px;
  margin: 1px;
  border: 3px solid #fff;
  border-radius: 50%;
  animation: lds-ring 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
  border-color: #fff transparent transparent transparent;
}


form {
  margin: 0 auto;
  width: 300px;
}

input {
  padding: 10px;
  width: 100%;
  border: 1px solid #CCC
}

button {
  padding: 10px
}

label {
  cursor: pointer
}

#form-switch {
  display: none
}

#register-form {
  display: none
}

#form-switch:checked~#register-form {
  display: block
}

#form-switch:checked~#login-form {
  display: none
}


main {
  display: flex;
  flex-direction: column;
  margin: auto;
  justify-content: center;
  align-items: center;
  height: 85vh;
}

.login--section {
  padding: 2rem;
  background-color: #ffffff;
  border-radius: .1rem;
  min-width: 310px;
}

.login--section__header {
  margin-bottom: 1.5rem;
  text-align: center;
  color: #3F45FF;
  font-weight: 600;
  font-size: 1.4rem;
}

.custom--form {
  font-weight: 400;
  font-size: 1rem;
  max-width: 100%;
}

.custom--form .form--group {
  position: relative;
  padding-bottom: 2px;
}

.submit--button {
  cursor: pointer;
  width: 100%;
  padding: .5rem;
  margin-bottom: 1rem;
  border-radius: .2rem;
  background-color: #3F45FF;
  color: #ffffff;
  border: #3F45FF;
  font-size: 1.2rem;
  transition: .3s ease all;
}

.submit--button:hover {
  border: #232994;
}

.instruction {
  display: flex;
}

.instruction__acount {
  color: #444444;
  margin-right: .5rem;
}

.last {
  margin-bottom: .2rem;
}

.instruction__login .instruction__signup {
  color: #3F45FF !important;
}

.signup--section {
  padding: .5rem 1rem 0;
  background-color: #ffffff;
}

.signup--section__header {
  margin: .5rem 0 0;
  text-align: center;
  color: #3F45FF;
  font-weight: 600;
  font-size: 2rem;
}

.custom--form {
  font-weight: 400;
  font-size: 1rem;
  max-width: 100%;
  margin-top: .4rem;
}

.custom--form .form--group .form--control {
  border-radius: .2rem;
  border-width: .1rem;
  height: auto;
  padding: .6rem;
  width: 100%;
}

.submit--button {
  cursor: pointer;
  width: 100%;
  padding: .5rem;
  margin-bottom: .3rem;
  border-radius: .2rem;
  background-color: #3F45FF;
  color: #ffffff;
  border: #3F45FF;
  font-size: 1.2rem;
  transition: .3s ease all;
}

.submit--button:hover {
  background-color: #232994;
}

.instruction {
  display: flex;
  margin-top: .8rem;
}

.instruction__acount {
  color: #444444;
  margin-right: .5rem;
}

.instruction__login .instruction__signup {
  color: #3F45FF;
}


@media screen and (min-width: 500px) {
  .main--nav {
    display: block;
  }
}
