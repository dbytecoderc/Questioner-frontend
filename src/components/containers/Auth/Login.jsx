import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import SimpleReactValidator from 'simple-react-validator';
import { Link, withRouter } from 'react-router-dom';

import Input from '../../presentationals/Input/Input';
import Button from '../../presentationals/Button/Button';
import Loader from '../../presentationals/Loader/Loader';

//Reducer
import { loginUser } from '../../../store/actions/auth';

import './Auth.css';
export class Login extends Component {
  state = {
    userDetails: {
      email: '',
      password: '',
    },
    errorMessage: '',
  };

  validator = new SimpleReactValidator({
    element: message => <div>{message}</div>,
  });

  onChange = event => {
    this.setState({
      userDetails: {
        ...this.state.userDetails,
        [event.target.name]: event.target.value,
      },
    });
  };
  formSubmitHandler = async event => {
    const { userDetails } = this.state;
    event.preventDefault();
    this.validator.showMessages();
    this.setState({
      userDetails: {
        ...this.state.userDetails,
      },
    });

    if (!this.validator.allValid()) {
      return false;
    }

    // let redirectUrl;

    // state ? (redirectUrl = state.from.pathname) : (redirectUrl = '/');

    const {
      history,
      // location: { state },
    } = this.props;

    await this.props.loginUser(this.state.userDetails, history);
    this.setState({
      errorMessage: this.props.auth.errorResponse[0],
    });
    this.setState({
      userDetails: {
        ...userDetails,
        password: '',
      },
    });
  };
  render() {
    const {
      auth: { isLoading },
    } = this.props;

    return (
      <div>
        <div id="login-form" className="login--section">
          <h1 className="login--section__header">Login</h1>
          <form onSubmit={this.formSubmitHandler} className="custom--form">
            <div className="form--group">
              <Input
                type="email"
                className="form--control"
                placeholder="E-mail"
                onChange={this.onChange}
                value={this.state.userDetails.email}
                name="email"
              />
            </div>
            <div className="validation--requirement">
              {this.validator.message(
                'email',
                this.state.userDetails.email,
                'required|email',
              )}
            </div>
            <div className="form--group">
              <Input
                type="password"
                className="form--control last"
                placeholder="Password"
                onChange={this.onChange}
                value={this.state.userDetails.password}
                name="password"
              />
            </div>
            <div className="validation--requirement">
              {this.validator.message(
                'password',
                this.state.userDetails.password,
                'required|min:6|alpha_num',
              )}
            </div>
            <div className="validation--requirement--last">
              {this.state.errorMessage}
            </div>

            <Button
              className="btn-dark submit--button"
              type="submit"
              value={isLoading ? <Loader /> : 'Login'}
            />
          </form>
          <div className="instruction">
            <p className="instruction__acount">Don't have an account?</p>
            <Link className="instruction__signup" to="/signup">
              Sign up
            </Link>
          </div>
        </div>
      </div>
    );
  }
}

Login.propTypes = {
  loginUser: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
};

const mapStateToProps = state => ({
  auth: state.auth,
  isLoading: state.auth.isLoading,
});

export default connect(
  mapStateToProps,
  { loginUser },
)(withRouter(Login));
