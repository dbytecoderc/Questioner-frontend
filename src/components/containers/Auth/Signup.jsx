import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import SimpleReactValidator from 'simple-react-validator';
import { Link, withRouter } from 'react-router-dom';

import Input from '../../presentationals/Input/Input';
import Button from '../../presentationals/Button/Button';
import Loader from '../../presentationals/Loader/Loader';

//Reducer
import { signupUser } from '../../../store/actions/auth';

import './Auth.css';

export class Signup extends Component {
  state = {
    userDetails: {
      firstname: '',
      email: '',
      username: '',
      password: '',
      confirmPassword: '',
    },
    errorMessage: '',
  };

  validator = new SimpleReactValidator({
    element: message => <div>{message}</div>,
    messages: {
      in: "The passwords don't match.",
    },
  });

  onChange = event => {
    this.setState({
      userDetails: {
        ...this.state.userDetails,
        [event.target.name]: event.target.value,
      },
    });
  };
  formSubmitHandler = async event => {
    event.preventDefault();
    this.validator.showMessages();
    this.setState({
      userDetails: {
        ...this.state.userDetails,
      },
    });

    if (!this.validator.allValid()) {
      return false;
    }

    let redirectUrl;

    state ? (redirectUrl = state.from.pathname) : (redirectUrl = '/');

    const {
      history,
      location: { state },
    } = this.props;

    await this.props.signupUser(this.state.userDetails, history, redirectUrl);
    this.setState({
      errorMessage: this.props.auth.errorResponse[0],
    });
  };

  render() {
    const { isLoading } = this.props.auth;
    return (
      <div>
        <div id="signup-form__wrapper" className="signup--section">
          <h1 className="signup--section__header">Sign up</h1>
          <form onSubmit={this.formSubmitHandler} className="custom--form">
            <div className="form--group">
              <Input
                type="text"
                className="form--control"
                placeholder="First name"
                onChange={this.onChange}
                value={this.state.userDetails.firstname}
                name="firstname"
              />
            </div>
            <div className="validation--requirement">
              {this.validator.message(
                'firstname',
                this.state.userDetails.firstname,
                'required|alpha_space',
              )}
            </div>
            <div className="form--group">
              <Input
                type="email"
                className="form--control"
                placeholder="Email"
                onChange={this.onChange}
                value={this.state.userDetails.email}
                name="email"
              />
            </div>
            <div className="validation--requirement">
              {this.validator.message(
                'email',
                this.state.userDetails.email,
                'required|email',
              )}
            </div>
            <div className="form--group">
              <Input
                type="text"
                className="form--control"
                placeholder="Username"
                onChange={this.onChange}
                value={this.state.userDetails.username}
                name="username"
              />
            </div>
            <div className="validation--requirement">
              {this.validator.message(
                'lastname',
                this.state.userDetails.username,
                'required|alpha_space',
              )}
            </div>
            <div className="form--group">
              <Input
                type="password"
                className="form--control"
                placeholder="Password"
                onChange={this.onChange}
                value={this.state.userDetails.password}
                name="password"
              />
            </div>
            <div className="validation--requirement">
              {this.validator.message(
                'password',
                this.state.userDetails.password,
                'required|min:6|alpha_num',
              )}
            </div>

            <div className="form--group">
              <Input
                type="password"
                className="form--control last"
                placeholder="Confirm Password"
                onChange={this.onChange}
                value={this.state.userDetails.confirmPassword}
                name="confirmPassword"
              />
              <div className="validation--requirement--last">
                {this.validator.message(
                  'confirmPassword',
                  this.state.userDetails.confirmPassword,
                  `required|in:${this.state.userDetails.password}`,
                )}
              </div>
              <div className="validation--requirement--last">
                {this.state.errorMessage}
              </div>
            </div>

            <Button
              className="submit--button"
              type="submit"
              value={isLoading ? <Loader /> : 'Sign up'}
            />
          </form>
          <div className="instruction">
            <p className="instruction__acount">Have an account?</p>
            <Link className="instruction__login" to="/login">
              Login
            </Link>
          </div>
        </div>
      </div>
    );
  }
}

Signup.propTypes = {
  signupUser: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
};

const mapStateToProps = state => ({ auth: state.auth });

export default connect(
  mapStateToProps,
  { signupUser },
)(withRouter(Signup));
