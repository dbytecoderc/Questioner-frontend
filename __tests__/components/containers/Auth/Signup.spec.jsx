import React from 'react';
import { shallow, mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';

import { Signup } from '../../../../src/components/containers/Auth/Signup';

const props = {
  auth: {
    isLoading: true,
    successResponse: {
      status: 'success',
      message: '',
    },
    errorResponse: [],
  },
  signupUser: jest.fn(),
};

const component = shallow(<Signup {...props} />);
const wrapper = mount(
  <BrowserRouter>
    <Signup {...props} />
  </BrowserRouter>,
);

describe('Signup component', () => {
  it('renders without errors', () => {
    expect(component).toMatchSnapshot();
  });

  it('should render a form with 2 input fields', () => {
    const inputs = wrapper.find('input').getElements();
    expect(inputs.length).toBe(5);
  });

  it('should correctly initialize the value of the email field to an empty string', () => {
    const input = wrapper.find('input[name="email"]');
    expect(input.prop('value')).toBe('');
  });

  it('should correctly initializes the value of the password field to an empty string', () => {
    const input = wrapper.find('input[name="password"]');
    expect(input.prop('value')).toBe('');
  });

  it('should correctly initialize the value of the email field to an empty string', () => {
    const input = wrapper.find('input[name="firstname"]');
    expect(input.prop('value')).toBe('');
  });

  it('should correctly initializes the value of the password field to an empty string', () => {
    const input = wrapper.find('input[name="username"]');
    expect(input.prop('value')).toBe('');
  });
  it('should correctly initialize the value of the email field to an empty string', () => {
    const input = wrapper.find('input[name="confirmPassword"]');
    expect(input.prop('value')).toBe('');
  });

  it('should not submit the form if there is an empty field', () => {
    const form = wrapper.find('form');
    form.simulate('submit', { preventDefault: jest.fn() });
    expect(wrapper.instance().state).toBe(null);
  });

  it('should submit the form and have an empty error body', () => {
    const email = wrapper.find('input[name="email"]');
    email.instance().value = 'igbominadeveloper@ah.com';
    email.simulate('change');
    const firstname = wrapper.find('input[name="firstname"]');
    firstname.instance().value = 'igbominadeveloper@ah.com';
    firstname.simulate('change');
    const username = wrapper.find('input[name="username"]');
    username.instance().value = 'igbominadeveloper@ah.com';
    username.simulate('change');
    const password = wrapper.find('input[name="password"]');
    password.instance().value = 'password1';
    password.simulate('change');
    const confirmPassword = wrapper.find('input[name="confirmPassword"]');
    confirmPassword.instance().value = 'password1';
    confirmPassword.simulate('change');
    const form = wrapper.find('form');
    form.simulate('submit', { preventDefault: jest.fn() });
  });
});
