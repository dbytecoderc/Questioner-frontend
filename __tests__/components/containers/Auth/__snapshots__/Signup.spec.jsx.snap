// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup component renders without errors 1`] = `
<div>
  <div
    className="signup--section"
    id="signup-form__wrapper"
  >
    <h1
      className="signup--section__header"
    >
      Sign up
    </h1>
    <form
      className="custom--form"
      onSubmit={[Function]}
    >
      <div
        className="form--group"
      >
        <Input
          className="form--control"
          name="firstname"
          onChange={[Function]}
          placeholder="First name"
          type="text"
          value=""
        />
      </div>
      <div
        className="validation--requirement"
      />
      <div
        className="form--group"
      >
        <Input
          className="form--control"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value=""
        />
      </div>
      <div
        className="validation--requirement"
      />
      <div
        className="form--group"
      >
        <Input
          className="form--control"
          name="username"
          onChange={[Function]}
          placeholder="Username"
          type="text"
          value=""
        />
      </div>
      <div
        className="validation--requirement"
      />
      <div
        className="form--group"
      >
        <Input
          className="form--control"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <div
        className="validation--requirement"
      />
      <div
        className="form--group"
      >
        <Input
          className="form--control last"
          name="confirmPassword"
          onChange={[Function]}
          placeholder="Confirm Password"
          type="password"
          value=""
        />
        <div
          className="validation--requirement--last"
        />
        <div
          className="validation--requirement--last"
        />
      </div>
      <Button
        className="submit--button"
        type="submit"
        value={<Loader />}
      />
    </form>
    <div
      className="instruction"
    >
      <p
        className="instruction__acount"
      >
        Have an account?
      </p>
      <Link
        className="instruction__login"
        to="/login"
      >
        Login
      </Link>
    </div>
  </div>
</div>
`;
